pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -113,6 +126,108 @@ func (p *provisioningAdmitter) Admit(request *admission.Request) (*admissionv1.A
 	return response, nil
 }
 
+// validateAgentEnvVars will validate the provisioning cluster object's AgentEnvVars, ensuring that the
+// "CATTLE_AGENT_VAR_DIR" env var is not allowed to be updated.
+func (p *provisioningAdmitter) validateAgentEnvVars(request *admission.Request, oldCluster, newCluster *v1.Cluster) *metav1.Status {
+	if newCluster.Spec.RKEConfig == nil {
+		return nil
+	}
+	if request.Operation == admissionv1.Create {
+		if slices.ContainsFunc(newCluster.Spec.AgentEnvVars, func(envVar rkev1.EnvVar) bool {
+			return envVar.Name == "CATTLE_AGENT_VAR_DIR"
+		}) {
+			return &metav1.Status{
+				Status:  "Failure",
+				Message: `"CATTLE_AGENT_VAR_DIR" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`,
+				Reason:  metav1.StatusReasonBadRequest,
+				Code:    http.StatusBadRequest,
+			}
+		}
+		return nil
+	}
+	if request.Operation == admissionv1.Update {
+		var oldCount, newCount int
+		for _, envVar := range oldCluster.Spec.AgentEnvVars {
+			if envVar.Name == "CATTLE_AGENT_VAR_DIR" {
+				oldCount++
+			}
+		}
+		for _, envVar := range newCluster.Spec.AgentEnvVars {
+			if envVar.Name == "CATTLE_AGENT_VAR_DIR" {
+				newCount++
+			}
+		}
+		if newCount > oldCount {
+			return &metav1.Status{
+				Status:  "Failure",
+				Message: `"CATTLE_AGENT_VAR_DIR" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`,
+				Reason:  metav1.StatusReasonBadRequest,
+				Code:    http.StatusBadRequest,
+			}
```

Comment:
Here and in several other places you are manually crafting the status field. This is a bit wordy (4 lines as opposed to 1) and doesn't explicitly set the `Result.Allowed` field. I would recommend using the webhook's `admission` package, which contains an `ResponseBadRequest` function [here](https://github.com/rancher/webhook/blob/276b2b953427d02f5ea2ae718ffa7e206399e725/pkg/admission/admission.go#L348)- this fills in a lot of these fields and sets the allowed flag on the response. You will need to return an `AdmissionResponse` object from your funciton, but I think that this will make it a lot more readable:

```suggestion
			return admission.ResponseBadRequest( `"CATTLE_AGENT_VAR_DIR" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`)
```